### PR TITLE
GC only wait for tensors generated by the main goroutine

### LINKED ISSUE #343: torch.GC may hang with a data loader cache
The current design requires all the tensors created between two consequent `torch.GC()` call being unreachable upon the second call. As a result, if we want to cache data loader tensors in a `chan`, the `torch.GC()` may hang forever. 

For a simplified example:
```golang
	torch.GC()
	runtime.LockOSThread()
	c := make(chan torch.Tensor, 0)
	{
		torch.NewTensor([][]float32{{1, 2}, {3, 4}}) // Register the anonymous `Tensor` to the global `WaitGroup`
		go func() {
			a := torch.NewTensor([][]float32{{1, 2}, {3, 4}}) // Register `a` to the global `WaitGroup`
			c <- a
			time.Sleep(time.Day)
			runtime.KeepAlive(&a)
		}()
	}
	<-c
	torch.GC() // Lasts for one day
```
The 2nd call to `torch.GC()` in the above code snippet will last for a day because `a` in the goroutine will live until the end of the goroutine.

### PR DESCRIPTION
**Pull request by shendiaomo**
Use `thread_local` and `runtime.LockOSThread()` together to make sure that `runtime.GC()` only waits for tensors created by the main goroutine to be reclaimed.

This is to enable the `ImageLoader` to do more work in the background goroutines.

Fix #343

### CODE REVIEW

File: cgotorch/memory.cc [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+#include "torch/torch.h"
+
+// FIXME(shendiaomo): including cgotorch.h before torch/torch.h will fail
+#include "cgotorch/cgotorch.h"
+
+thread_local bool gcPrepared;
+
+uint8_t GCPrepared() { return gcPrepared; }
+
+void PrepareGC() { gcPrepared = true; }
```
Comment (wangkuiyi) [inline cgotorch/memory.cc:11]: Do we need to make sure that the assignment is atomic?

File: tensor_gc.go [hunk 2]
```diff
@@ -13,20 +13,19 @@ import (
...
-	gcPrepared         = false
 )
 
 // SetTensorFinalizer sets a finalizer to the tensor
 func SetTensorFinalizer(t *unsafe.Pointer) {
 	// We don't want the following conditional and the finalizer using
 	// different gcPrepared values, so we leverage p and closure here.
-	p := gcPrepared
-	if p {
+	p := C.GCPrepared()
```
Comment (wangkuiyi) [inline tensor_gc.go:22]: Do we need a lock when calling C.GCPrepared() or in that function?

File: tensor_test.go [hunk 3]
```diff
@@ -123,3 +125,21 @@ func TestTensorPinMemory(t *testing.T) {
...
+		torch.NewTensor([][]float32{{1, 2}, {3, 4}})
+		go func() {
+			a := torch.NewTensor([][]float32{{1, 2}, {3, 4}})
+			c <- true
+			time.Sleep(time.Second)
+			runtime.KeepAlive(&a)
+		}()
+	}
+	<-c
+	assert.Eventually(t, func() bool { torch.GC(); return true }, 2*time.Millisecond, 10*time.Microsecond)
```
Comment (wangkuiyi) [inline tensor_test.go:143]: Why do we need to call Eventually here? With the `<- c`, it seems that we can directly call `torch.GC()` here.

File: tensor_test.go [hunk 4]
```diff
@@ -123,3 +125,21 @@ func TestTensorPinMemory(t *testing.T) {
...
+	torch.GC()
+	runtime.LockOSThread()
+	c := make(chan bool, 0)
+	{
+		torch.NewTensor([][]float32{{1, 2}, {3, 4}})
+		go func() {
+			a := torch.NewTensor([][]float32{{1, 2}, {3, 4}})
+			c <- true
+			time.Sleep(time.Second)
+			runtime.KeepAlive(&a)
```
Comment (wangkuiyi) [inline tensor_test.go:139]: Is the purpose of calling `KeepAlive` to check that `torch.GC` will free up two tensors -- on created in L134 and the one in L136?  However, I  don't see a checking in this test case for the freeing up of two tensors?

**Review by wangkuiyi — COMMENTED** [review]
I don't understand this change -- the purpose and the solution. Need more explanation.

File: cgotorch/memory.cc (hunk 1 above)
Comment (shendiaomo) [reply cgotorch/memory.cc:11]: No, we don't. `gcPrepared` is a `thread_local`, each thread has its only copy.

File: tensor_gc.go (hunk 2 above)
Comment (shendiaomo) [reply tensor_gc.go:22]: No, as explained above. `C.GCPrepared` returns a copy of the`thread_local`.

File: tensor_test.go (hunk 3 above)
Comment (shendiaomo) [reply tensor_test.go:143]: We call `Eventually` here to make sure that `torch.GC()` doesn't have to wait the `Tensor a` in the goroutine to be unreachable.

File: tensor_test.go (hunk 4 above)
Comment (shendiaomo) [reply tensor_test.go:139]: This is to tell the escape analysis of Go garbage collector not to mark `a` as unreachable before the call of `KeepAlive`. 

Without the `thread_local`, `torch.GC()` will recycle the `Tensor` in L134 and hang until the goroutine ends, and then recycle the `Tensor` in L136. 

In practice, without this fix, if we keep a cache and fill it in a goroutine, calling `torch.GC` in the main goroutine will certainly hang the training process forever.

File: tensor_test.go [hunk 5]
```diff
@@ -123,3 +125,21 @@ func TestTensorPinMemory(t *testing.T) {
 		assert.Equal(t, " 1  2\n 3  4\n[ CPUFloatType{2,2} ]", b.String())
 	}
 }
+
+func TestTensorGC(t *testing.T) {
+	torch.GC()
+	runtime.LockOSThread()
+	c := make(chan bool, 0)
```
Comment (Yancey1989) [inline tensor_test.go:132]: To simulate `ImageLoader`  workload, maybe we shoul use the Tensor channel as the following code: 

``` golang
c := make(chan Tensor)
go func() {
    a := torch.NewTensor([][]float32{{1, 2}, {3, 4}})
    c <- a
    time.Sleep(time.Second)
    runtime.KeepAlive(&a)
}()
a := <- c
```

Comment (shendiaomo) [reply tensor_test.go]: Done.

File: tensor_test.go [hunk 6]
```diff
@@ -144,3 +146,20 @@ func TestTensorPinMemory(t *testing.T) {
 		assert.Equal(t, " 1  2\n 3  4\n[ CPUFloatType{2,2} ]", b.String())
 	}
 }
+
+func TestTensorGC(t *testing.T) {
+	torch.GC()
+	runtime.LockOSThread()
```
Comment (Yancey1989) [inline tensor_test.go:152]: We have called `runtime.LocalOSTread()` function at  https://github.com/wangkuiyi/gotorch/blob/develop/device.go#L89 , can we remove this line?

File: tensor_test.go [hunk 7]
```diff
@@ -144,3 +146,20 @@ func TestTensorPinMemory(t *testing.T) {
 		assert.Equal(t, " 1  2\n 3  4\n[ CPUFloatType{2,2} ]", b.String())
 	}
 }
+
+func TestTensorGC(t *testing.T) {
+	torch.GC()
```
Comment (Yancey1989) [inline tensor_test.go:151]: `defer torch.FinishGC()`

File: tensor_test.go (hunk 6 above)
Comment (shendiaomo) [reply tensor_test.go:152]: No. `go test` runs test cases in separate goroutines. The call of `runtime.LocalOSThread` only locks the *main goroutine*.

File: tensor_test.go (hunk 7 above)
Comment (shendiaomo) [reply tensor_test.go:151]: Done.

**Review by Yancey1989 — APPROVED** [review]
LGTM!

**Pull request merged**